### ISSUE TITLE
Do not modify global random state

### ISSUE THREAD

**Issue by mitar** [pytorch/pytorch#39716]
## 🚀 Feature

Currently, the [recommended approach to achieve reproducibility](https://pytorch.org/docs/stable/notes/randomness.html) is setting global random seeds. I would like to propose that instead all functions which need a random source accept a local, non-global, `random_seed`/`random_state` argument to which one can pass a random state object, or numpy generator, or something else, which binds random source locally.

## Motivation

Reproducibility if often needed in ML programs. But current approach of setting global random seed in various places in pytorch is suitable only for small programs which are using only pytorch. Now imagine that you have many libraries around, each trying to set global random seed at various times. This can lead to strange changes in random seeds just because you swapped two calls around. Moreover, if any of those libraries stand-alone produces a sequence of numbers, it might be that when used in combination with pytorch the sequence is not anymore the same. So if you are trying to compare two different implementations (one in pytorch one in some other library), the behavior of the implementation in the other library can change just because you also have pytorch around.

Of course, the other library should also not be using global random source. But ideally pytorch should not either.

In our concrete case, we have an AutoML system generating many candidate pipelines using primitives to build those pipelines which in turn use many different libraries. Those pipelines are run and we hope to have full reproducibility (which helps with metalearning). But setting/resetting global random seed just because one primitive in one candidate pipeline we are exploring happens to use pytorch influences behavior of unrelated pipelines.

## Pitch

Instead of relying on global random sources, add `random_state` parameter to all functions which expect/use randomness. Do not ever set random seed of a global random source in pytorch code (the user of pytorch can do that themselves if they really want).

## Alternatives

Use context managers to set and reset random seeds, but this becomes tricky when using multi-threading.

## Additional context

See example in [sklearn](https://scikit-learn.org/stable/glossary.html#term-random-state).

**Comment by ShawnZhong** [comment]
I think they already exist in PyTorch?

```py
with torch.random.fork_rng(devices=rng_device):
    torch.some_rand_func(some_input)
```
```py
torch.some_rand_func(some_input, generator=torch.Generator().manual_seed(42))
```

**Comment by mitar** [comment]
So using a context manager is still modifying the global random seed, and is discussed under "Alternatives".

The doc I linked above, suggests the use of `torch.manual_seed` for reproducibility. This is modifying global random source. It also suggests setting random seed on numpy. Moreover, [every worker](https://github.com/pytorch/pytorch/blob/78d5707041ea4c3f9e5e8aa13f6858d54de5a6da/torch/utils/data/_utils/worker.py#L121-L123) sets global random seeds.

**Comment by albanD** [comment]
Hi,

All the rng functions accept a `generator=` parameter that will be used instead of the global one.
So if you want to avoid global rng state, you can simply provide the generator for each rng op that you do.
Does that fit your needs?

**Comment by mitar** [comment]
So [this codepath does not seem to have this option](https://github.com/pytorch/pytorch/blob/78d5707041ea4c3f9e5e8aa13f6858d54de5a6da/torch/utils/data/_utils/worker.py#L121-L123).

**Comment by mitar** [comment]
Maybe this issue is just a duplicate of #39572 it seems, and of an issue that documentation could be improved and `generator=` suggested as recommended way (when you control all code) and setting global seed as a fallback if you do not control pytorch code. I could do a PR to update documentation in this way, if that is agreeable?

**Comment by albanD** [comment]
> So this codepath does not seem to have this option.

Well the thing is that if you set the generator, you can ignore this code and the fact that it sets a global state because you don't use that global state anyways :D 
The performance penalty from doing this is going to be zero compared to the process initialization.

If your concern is more on how to use these generators with the builtin dataloader sampler, then I think this is a duplicate of #39572 yes.

**Comment by mitar** [comment]
> Well the thing is that if you set the generator, you can ignore this code and the fact that it sets a global state because you don't use that global state anyways

You are right. In those workers there is no other code really running.

About about my suggestion on improving documentation?

**Comment by albanD** [comment]
I think we can add a comment about it in the reproducibility note but I'm not sure we want to make it the recommended way.
The global state (even though it has shortcomings) should be enough for most people and is much simpler to use.

**Comment by SsnL** [comment]
I don't think using a generator object should be the recommended way. Yes, it makes combining code less error prune, but it makes code a lot longer and won't affect most use cases anyways. I do agree with adding a note.

**Comment by RuRo** [comment]
Currently, the fact that `pytorch` and `numpy` use per-process global random state means that you can't use threading parallelism for running reproducible experiments, which is a huge pain, since full process parallelism is harder to get right and has significant overhead due to pickling.

Consider the following example:
```python
import time
import concurrent.futures as fut

import torch
import numpy as np

def randint_with_seed(seed):
    np.random.seed(seed)
    torch.manual_seed(seed)

    time.sleep(1)

    return (
        np.random.randint(0, 2**32, size=(1,)).item(),
        torch.randint(0, 2**32, size=(1,)).numpy().item(),
    )

print("Expected outputs")
for _ in range(4):
    for seed in range(4):
        print(randint_with_seed(seed))
    print()

print("With multithreading")
with fut.ThreadPoolExecutor() as pool:
    for _ in range(4):
        tasks = []
        for seed in range(4):
            tasks.append(pool.submit(randint_with_seed, seed))
        for task in tasks:
            print(task.result())
        print()
```

The `randint_with_seed` function simulates a deterministic experiment. As you can see, the sequential run is deterministic, while the parallel run gives different results each time.

**Comment by rsokl** [comment]
Given that [numpy has completely revised their PRNG API](https://numpy.org/devdocs/reference/random/index.html) to avoid global random state and [jax avoids mutable RNG objects altogether](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers), I wish PyTorch would take a more active approach toward dissuading users from setting a global random seed.

The comments above – that the current recommendation is "good enough for most people" and that passing around Generators "won't effect most uses cases anyways" – seem poorly motivated to me.  The documentation for achieving reproducibility should show people how to make randomness as local and "non-spooky" as possible, not "here is how to set up a Rube Goldberg machine that will work the same way each time, as long as you don't change a single thing". Library authors look to this document to understand how they should ensure that their library enables reproducibility, and they get essentially no real guidance.
